Sprint 17

ArtemDubovitskiy opened the pull request:
Прошу проверить самостоятельную задачу по 17 спринту:
 - Добавлено Контекстное меню, фильтры
 - Добавлен экран статистики
 - Главный экран локализован (ru / en)
 - Реализована темная тема и скриншот тесты
 - Подключена аналитика (AppMetrica)
 
Спасибо!)

ArtemDubovitskiy committed:
Добавлено:
- исправлено сохранение категории в СoreData на экране создания категории;
- добавлена проверка на созданную категорию и отображение алерта (если категория уже создана)
- поправлен Алерт (добавлено изменение названия);

ArtemDubovitskiy committed:
Добавлено:
- контекстное меню на экране трекеров
- исправлены констрейнты для collectionView для правильного отображения контекстного меню
- исправлена ячейка collectionView для правильной работы контекстного меню

ArtemDubovitskiy committed:
Добавлено:
- закрепление трекера в категорию "Закрепленные"

ArtemDubovitskiy committed:
Добавлено удаление трекера

ArtemDubovitskiy committed:
Добавлено удаление категории

ArtemDubovitskiy committed:
Добавлено редактирование трекера

ArtemDubovitskiy committed:
Добавлено:
- кнопка фильтров на экран трекеров
- filterViewController
- отображение булавки на карточке трекера когда он закреплен

ArtemDubovitskiy committed:
Доделан экран статистики с получением данных из coreData

ArtemDubovitskiy committed:
Добалено редактирование категорий

ArtemDubovitskiy committed:
Добавлена локализация главного экрана

ArtemDubovitskiy committed:
Добавлены скриншот тесты для светлой темы

ArtemDubovitskiy committed:
Добавлены скриншот тесты

ArtemDubovitskiy committed:
Добавлена Аналитика (AppMetrica)

ArtemDubovitskiy committed:
Добавлена темная тема

ArtemDubovitskiy committed:
Поправлена темная тема, исправлен баг с запуском приложения

ArtemDubovitskiy committed:
Исправлено:
- темная тема онбординга,
- локализация дней на экране редактирования трекера

So1ovyov commented on Tracker/Statistics/StaticticsViewController.swift:
```diff
@@ -59,7 +82,88 @@ final class StaticticsViewController: UIViewController {
+        
+        switch indexPath.row {
+        case 0:
+            count = "0"
```
Если у тебя тут отсутствует логика, может стоит добавить TODO что это моковые данные (просто ноль) и в следующей доработке добавить отображение статистики по другим ячейкам

So1ovyov commented on Tracker/Statistics/StatisticsCell.swift:
```diff
@@ -0,0 +1,95 @@
+        backgroundColor = .clear
+        clipsToBounds = true
+        
+        addSubview(borderView)
```
Обрати внимание на инструмент `forEach` при верстке и настройке элементов, он поможет тебе сократить повторяющийся код
Вот пример:
```
        [view1, 
         view2,
         view3].forEach {
            $0.translatesAutoresizingMaskIntoConstraints = false
             view.addSubview($0)
        }
```
Таким образом все строки `translatesAutoresizingMaskIntoConstraints` и `addSubview` выше можно удалить и очень сильно сократить код

ArtemDubovitskiy replied:
спасибо!

So1ovyov commented on Tracker/Trackers/Category/CategoryViewController.swift:
```diff
@@ -170,6 +172,51 @@ extension CategoryViewController: UITableViewDelegate {
+                   point: CGPoint) -> UIContextMenuConfiguration? {
+        let category = self.categoryViewModel.categories[indexPath.row]
+        
+        let configuration = UIContextMenuConfiguration(identifier: nil,
```
Очень много логики в методе tableView, может ее вынести в отдельный private метод ?

ArtemDubovitskiy replied:
да, я планирую это сделать далее (когда буду все приложение переписывать на MVVM)

So1ovyov commented on Tracker/Trackers/Category/CategoryViewModel.swift:
```diff
@@ -26,17 +26,37 @@ final class CategoryViewModel {
+        do {
+            try self.trackerCategoryStore.deleteCategory(category)
+        } catch {
+            print("Error delete category: \(error.localizedDescription)")
```
Было бы также круто добавить TODO "Обработать ошибку"

So1ovyov commented on Tracker/Trackers/Category/CreateCategoryViewController.swift:
```diff
@@ -7,28 +7,32 @@
 final class CreateCategoryViewController: UIViewController {
     weak var delegate: CreateCategoryViewControllerDelegate?
+    var existingCategory: TrackerCategory?
+    private (set) var categoryViewModel = CategoryViewModel()
```
Почему для viewModel ты используешь `private (set)` может имеет смысл сделать просто приватным это свойство ?

So1ovyov commented on Tracker/Trackers/Filter/FilterViewController.swift:
```diff
@@ -0,0 +1,157 @@
+final class FilterViewController: UIViewController {
+    weak var delegate: FilterViewControllerDelegate?
+    static var filterIndex: Int = 0
+    var trackerFilters = ["Все трекеры",
```
Точно ли этот массив должен торчать наружу ?

ArtemDubovitskiy replied:
Спасибо, поправил)

So1ovyov commented on Tracker/Trackers/TrackerCollectionViewCell.swift:
```diff
@@ -101,13 +110,9 @@ final class TrackerCollectionViewCell: UICollectionViewCell {
-    
     @objc
     private func plusTrackerButtonTapped() {
+        analyticsService.report(event: "click", params: ["screen": "Main", "item": "track"])
```
Во всем проекте только одно событие аналитики, кажется по ТЗ не хватает еще других событий

So1ovyov commented on Tracker/Trackers/TrackersViewController.swift:
```diff
@@ -386,6 +528,119 @@ extension TrackersViewController: UICollectionViewDataSource {
+        let tracker = self.visibleCategories[indexPath.section].trackers[indexPath.row]
+        let index = "\(indexPath.row):\(indexPath.section)" as NSString
+        
+        let configuration = UIContextMenuConfiguration(identifier: index,
```
Очень большой метод, давай его декомпозируем и вынесем часть логики в отдельный приватный метод. Можно попробовать создать фабрику для создания `UIContextMenu`

ArtemDubovitskiy replied:
да, я планирую это сделать далее (когда буду все приложение переписывать на MVVM)

ArtemDubovitskiy committed:
Ипсралено по замечаниям ревью:
Закрепление / Открепление трекера:
- исправлено закрепление / открепление трекера в collectionView.
- добавлено отображение плейсхолдеров при обновленном состоянии
Скриншот тесты:
- добавлен скриншот-тест для темной темы
- исправлен цвет фона collectionView (из-за которого не проходили скриншот тесты)
- исправлено отображение плейсхолдеров
- скриншот тесты выполнены на устройстве с английским языком
- в тестах добавлена возможность изменить фон / при изменении тест падает (как и должно быть)
Локализация:
- добавлена локализация en/UK.
Проверено на симуляторе: Iphone 14/15; регион РФ, UK , USA
Проверено на iphone 11/13: региоон РФ, русский и английский язык интерфейса

ArtemDubovitskiy committed:
Исправлено отображение трекеров после редактировая

ArtemDubovitskiy committed:
Исправлено по замечаниям ревью:
- добавлна фильтрация для выполненных и невыполнынных трекеров.
